Zend/zend_API.h

```diff
@@ -674,6 +677,120 @@ END_EXTERN_C()
 #define ZEND_GINIT_FUNCTION			ZEND_MODULE_GLOBALS_CTOR_D
 #define ZEND_GSHUTDOWN_FUNCTION		ZEND_MODULE_GLOBALS_DTOR_D
 
+static zend_always_inline int zend_try_assign_ex(zval *zv, zval *arg, zend_bool strict) {
+	zend_reference *ref;
+	if (EXPECTED(Z_ISREF_P(zv)) && UNEXPECTED(ZEND_REF_HAS_TYPE_SOURCES(ref = Z_REF_P(zv)))) {
+		zval tmp;
+		ZVAL_COPY_VALUE(&tmp, arg);
+		if (EXPECTED(zend_verify_ref_assignable_zval(ref, &tmp, strict))) {
+			zv = Z_REFVAL_P(zv);
+			zval_ptr_dtor(zv);
+			ZVAL_COPY_VALUE(zv, &tmp);
+		} else {
+			zval_ptr_dtor(&tmp);
+			return FAILURE;
+		}
+	} else {
+		ZVAL_DEREF(zv);
+		zval_ptr_dtor(zv);
+		ZVAL_COPY_VALUE(zv, arg);
+	}
+
+	return SUCCESS;
+} 
+
+static zend_always_inline int zend_try_assign(zval *zv, zval *arg) {
+	return zend_try_assign_ex(zv, arg, ZEND_ARG_USES_STRICT_TYPES());
```

Comment:
I agree here, the zend_parse_arg_* APIs suffer from the same issue as well; as far as I understood it, it's to keep the API smaller/cleaner.
Feel free to suggest a concrete path for this while still keeping a simple API. (ZEND_TRY_ASSIGN_*() macros)

However we do not need these functions _that_ often, there are not _that_ many functions writing to by-ref args, that it would have any noticeable impact...